endpoints/s3/src/putobject.cpp

```diff
@@ -18,18 +18,57 @@
 
 #include <boost/beast/http/empty_body.hpp>
 #include <boost/beast/http/read.hpp>
+#include <boost/lexical_cast.hpp>
 
 #include <iostream>
 #include <vector>
 #include <fstream>
 #include <regex>
 #include <memory>
+#include <unordered_map>
 
 namespace asio = boost::asio;
 namespace beast = boost::beast;
 namespace fs = irods::experimental::filesystem;
 namespace logging = irods::http::logging;
 
+namespace irods::s3::api::part_shared_memory
+{
+	// This is a part size map which is shared between threads. Whenever part sizes are known
+	// they are added to this map. When complete or cancel multipart is executed, the part sizes
+	// for that upload_id are removed.
+	// The map looks like the following with the key of the first map being the upload_id and the
+	// key of the second map being the part number:
+	//    {
+	//      "1234abcd-1234-1234-123456789abc":
+	//        { 0: 4096000,
+	//          1: 4096000,
+	//          4: 4096000 },
+	//      "01234abc-0123-0123-0123456789ab":
+	//        { 0: 5192000,
+	//          3: 5192000 }
+	//    }
+	std::unordered_map<std::string, std::unordered_map<unsigned int, uint64_t>> part_size_map;
+
+	// This map holds persistent data needed for each upload_id. This includes the replica_token and
+	// replica_number.  In addition it holds shared pointers for the connection, transport, and odstream
+	// for the first stream that is opened. These shared pointers are saved to this map so that the first
+	// open() to an iRODS data object can remain open throughout the lifecycle of the request.  This
+	// stream will be closed in either CompleteMultipartUpload or AbortMultipartUpload.
+	std::unordered_map<
+		std::string,
+		std::tuple<
+			irods::experimental::io::replica_token,
+			irods::experimental::io::replica_number,
+			std::shared_ptr<irods::experimental::client_connection>,
+			std::shared_ptr<irods::experimental::io::client::native_transport>,
+			std::shared_ptr<irods::experimental::io::odstream>>>
```

Comment:
Consider adding some type aliases for things like this and even the type for `part_size_map`. Would warrant a Great Naming...

Reply:
I am going to hold off on the alias until we can discuss naming.

Reply:
Please open an issue for this and link it to this conversation.

We can address this in a later release.

Reply:
Issue opened.

Reply:
https://github.com/irods/irods_client_s3_api/issues/123

Reply:
Issue number is #123.